### PR TITLE
[BugFix] Resolve DDP hanging bug within ModelCheckpoint and val_loss

### LINKED ISSUE #5865: Training stuck at 0% after few epochs while training with DDP
## 🐛 Bug

I recently updated to pytorch_lightning 1.1.7 and noticed that after a few epochs of training, the training % is stuck at 0% and never progresses. When I switched back to 1.1.4, this strange behavior does not occur. I do not know the root cause of this issue.

 - PyTorch Version (e.g., 1.0): 1.1.7
 - OS (e.g., Linux): Linux, U18
 - How you installed PyTorch (`conda`, `pip`, source): pip install 
 - Build command you used (if compiling from source):
 - Python version: 3.6
 - CUDA/cuDNN version: 10.0
 - GPU models and configuration: RTX 2080 x3
 - Any other relevant information:

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Currently, the training can hang under the following conditions:
- No `monitor` is being provided to ModelCheckPoint
- val_loss is being logged

Internally in ModelCheckpoint, self.monitor will become val_loss.
```
    def _add_backward_monitor_support(self, trainer):
        metrics = trainer.logger_connector.callback_metrics
        # backward compatibility... need to deprecate
        if self.monitor is None and 'val_loss' in metrics:
            self.monitor = 'val_loss'
        if self.monitor is None and 'checkpoint_on' in metrics:
            self.monitor = 'checkpoint_on'
        if self.save_top_k is None and self.monitor is not None:
            self.save_top_k = 1
```
And the code will diverge slightly later if check_monitor_top_k returns different value for both processes. 

```
self.check_monitor_top_k(metrics.get(self.monitor))
```

Solution:

- `val_loss` and `checkpoint_on` should be reduced.

Fixes #5865

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -548,6 +548,12 @@ def _save_top_k_checkpoints(self, trainer, pl_module, metrics):
         epoch = metrics.get("epoch")
         step = metrics.get("step")
 
+        # when `val_loss` is being logged and no ModelCheckpoint is being provided
+        # `val_loss` or `checkpoint_on` will be selected for monitor and need to be reduced to
+        # prevent processes divergence
+        if self.monitor in ("val_loss", "checkpoint_on"):
```
Comment (justusschock) [inline pytorch_lightning/callbacks/model_checkpoint.py:554]: Shouldn't we reduce it always?

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/callbacks/model_checkpoint.py:554]: Metrics already perform reduction during compute. I think the issue only happens with legacy metrics.

Comment (awaelchli) [reply pytorch_lightning/callbacks/model_checkpoint.py:554]: I'm not sure about this. I thought the metrics that the checkpoint gets from the trainer/logger connector are already reduced? We shouldn't let the checkpoint have the responsibility to reduce metrics or to assume how. the mean is not always correct.

Comment (SeanNaren) [reply pytorch_lightning/callbacks/model_checkpoint.py:554]: can we get a test case of this legacy metric? It will help debug and figure out why this reduction is necessary

**Comment by carmocca** [comment]
Note that we should deprecate and remove `_add_backward_monitor_support` (we never got around to do it). I can do it in another PR

**Comment by tchaton** [comment]
> Note that we should deprecate and remove `_add_backward_monitor_support` (we never got around to do it). I can do it in another PR

Sounds good !

**Comment by SeanNaren** [comment]
Seems like this didn't fix the issue in #5865. We're really missing a test case here for us to debug, so might be a good idea to get the user to help out here.
